tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -386,188 +398,956 @@ impl ServerKey {
+                        // block that encodes the 2 values needed to later know if overflow did
+                        // happen depending on the input carry of the last block.
+                        let lut = self.key.generate_lookup_table_bivariate(|lhs, rhs| {
+                            let mask = (1 << (num_bits_in_message - 1)) - 1;
+                            let lhs_except_last_bit = lhs & mask;
+                            let rhs_except_last_bit = rhs & mask;
+
+                            let overflows_with_given_input_carry = |input_carry| {
+                                let output_carry =
+                                    ((lhs + rhs + input_carry) >> num_bits_in_message) & 1;
+
+                                let input_carry_to_last_bit =
+                                    ((lhs_except_last_bit + rhs_except_last_bit + input_carry)
+                                        >> (num_bits_in_message - 1))
+                                        & 1;
+
+                                u64::from(input_carry_to_last_bit != output_carry)
+                            };
+
+                            (overflows_with_given_input_carry(1) << 3)
+                                | (overflows_with_given_input_carry(0) << 2)
+                        });
+                        let (last_lhs_block, last_rhs_block) = saved_last_blocks.as_ref().unwrap();
+                        self.key.unchecked_apply_lookup_table_bivariate(
+                            last_lhs_block,
+                            last_rhs_block,
+                            &lut,
+                        )
+                    },
+                    || {
+                        let (shifted_blocks, mut block_states) =
+                            self.compute_shifted_blocks_and_block_states(blocks);
+                        let _ = block_states.pop().unwrap();
+                        (shifted_blocks, block_states)
+                    },
+                );
+
+                output_flag = Some(block);
+                (shifted_blocks, block_states)
+            }
+            OutputFlag::Carry => {
+                let (shifted_blocks, mut block_states) =
+                    self.compute_shifted_blocks_and_block_states(blocks);
+                let last_block_state = block_states.pop().unwrap();
+                output_flag = Some(last_block_state);
+                (shifted_blocks, block_states)
+            }
+        };
+
+        // Second step
+        let (mut prepared_blocks, mut groupings_pgns) = {
+            // This stores, the LUTs that given a cum sum block in the first grouping
+            // tells if a carry is generated or not
+            let first_grouping_inner_propagation_luts = (0..grouping_size - 1)
+                .map(|index| {
+                    self.key.generate_lookup_table(|propa_cum_sum_block| {
+                        let carry = propa_cum_sum_block & (1 << index);
```

Comment:
could we keep a consistent style for those bit reads ? I think shifting down and doing & 1 reads a bit better and is equivalent ?

it's a nitpick so if this is better for a reason I don't see fine to keep